File: capture_bytecode.lua
-- Bytecode Capture Script for Luraph VM
-- Captures decrypted bytecode by hooking the VM execution

-- Lua 5.4 bit32 compatibility layer
bit32 = {
    band = function(a, b, ...)
        a = a or 0
        b = b or 0
        local result = a & b
        for i = 1, select('#', ...) do
            local v = select(i, ...) or 0
            result = result & v
        end
        return result
    end,
    bor = function(a, b, ...)
        a = a or 0
        b = b or 0
        local result = a | b
        for i = 1, select('#', ...) do
            local v = select(i, ...) or 0
            result = result | v
        end
        return result
    end,
    bxor = function(a, b, ...)
        a = a or 0
        b = b or 0
        local result = a ~ b
        for i = 1, select('#', ...) do
            local v = select(i, ...) or 0
            result = result ~ v
        end
        return result
    end,
    bnot = function(a)
        a = a or 0
        return ~a & 0xFFFFFFFF
    end,
    lshift = function(a, b)
        a = a or 0
        b = b or 0
        return (a << b) & 0xFFFFFFFF
    end,
    rshift = function(a, b)
        a = a or 0
        b = b or 0
        return (a & 0xFFFFFFFF) >> b
    end,
    arshift = function(a, b)
        a = (a or 0) & 0xFFFFFFFF
        b = b or 0
        if a >= 0x80000000 then
            return ((a >> b) | (~(0xFFFFFFFF >> b))) & 0xFFFFFFFF
        end
        return a >> b
    end,
    lrotate = function(a, b)
        a = (a or 0) & 0xFFFFFFFF
        b = (b or 0) % 32
        return ((a << b) | (a >> (32 - b))) & 0xFFFFFFFF
    end,
    rrotate = function(a, b)
        a = (a or 0) & 0xFFFFFFFF
        b = (b or 0) % 32
        return ((a >> b) | (a << (32 - b))) & 0xFFFFFFFF
    end,
    extract = function(a, field, width)
        a = a or 0
        field = field or 0
        width = width or 1
        return (a >> field) & ((1 << width) - 1)
    end,
    replace = function(a, v, field, width)
        a = a or 0
        v = v or 0
        field = field or 0
        width = width or 1
        local mask = (1 << width) - 1
        return (a & ~(mask << field)) | ((v & mask) << field)
    end,
    btest = function(...)
        return bit32.band(...) ~= 0
    end,
    countlz = function(a)
        a = (a or 0) & 0xFFFFFFFF
        if a == 0 then return 32 end
        local n = 0
        if a <= 0x0000FFFF then n = n + 16; a = a << 16 end
        if a <= 0x00FFFFFF then n = n + 8; a = a << 8 end
        if a <= 0x0FFFFFFF then n = n + 4; a = a << 4 end
        if a <= 0x3FFFFFFF then n = n + 2; a = a << 2 end
        if a <= 0x7FFFFFFF then n = n + 1 end
        return n
    end,
    countrz = function(a)
        a = (a or 0) & 0xFFFFFFFF
        if a == 0 then return 32 end
        local n = 0
        if (a & 0x0000FFFF) == 0 then n = n + 16; a = a >> 16 end
        if (a & 0x000000FF) == 0 then n = n + 8; a = a >> 8 end
        if (a & 0x0000000F) == 0 then n = n + 4; a = a >> 4 end
        if (a & 0x00000003) == 0 then n = n + 2; a = a >> 2 end
        if (a & 0x00000001) == 0 then n = n + 1 end
        return n
    end
}

-- Luau compatibility
if not getgenv then
    getgenv = function() return _G end
end

-- Wrap string.pack to handle negative numbers for unsigned formats
local original_string_pack = string.pack
string.pack = function(fmt, ...)
    local args = {...}
    local new_args = {}

    -- Convert all numeric arguments that might be negative
    for i, v in ipairs(args) do
        if type(v) == "number" and v < 0 then
            -- Convert to unsigned 32-bit equivalent
            new_args[i] = math.floor(v) & 0xFFFFFFFF
        else
            new_args[i] = v
        end
    end

    local ok, result = pcall(original_string_pack, fmt, table.unpack(new_args))
    if ok then
        return result
    else
        -- If still failing, try to be more aggressive
        for i, v in ipairs(new_args) do
            if type(v) == "number" then
                new_args[i] = math.floor(v) % (2^32)
            end
        end
        return original_string_pack(fmt, table.unpack(new_args))
    end
end

-- Output file
local output = io.open("captured_bytecode.lua", "w")
local function log(msg)
    output:write(msg .. "\n")
    output:flush()
    print(msg)
end

log("-- Luraph VM Bytecode Capture")
log("-- " .. os.date())
log("")

-- Function counter
local func_count = 0

-- Helper to dump arrays
local function dump_array(arr, name, max_len)
    max_len = max_len or 5000
    if type(arr) ~= "table" then
        log(name .. " = " .. tostring(arr))
        return
    end

    local entries = {}
    for i = 1, max_len do
        if arr[i] ~= nil then
            local v = arr[i]
            local vs
            if type(v) == "string" then
                local s = v:sub(1, 100)
                s = s:gsub("\\", "\\\\"):gsub("\n", "\\n"):gsub("\r", "\\r"):gsub('"', '\\"')
                s = s:gsub("[%c]", function(c) return string.format("\\x%02X", string.byte(c)) end)
                vs = string.format('"%s"', s)
                if #v > 100 then vs = vs .. " --[[" .. #v .. " chars]]" end
            elseif type(v) == "number" then
                vs = tostring(v)
            elseif type(v) == "boolean" then
                vs = tostring(v)
            elseif type(v) == "function" then
                vs = "<function>"
            elseif type(v) == "table" then
                vs = "<table:" .. #v .. ">"
            else
                vs = "<" .. type(v) .. ">"
            end
            table.insert(entries, string.format("  [%d] = %s,", i, vs))
        end
    end

    log(name .. " = {")
    for _, e in ipairs(entries) do
        log(e)
    end
    log("}")
    log("")
end

-- Load the VM
log("-- Loading VM code...")
local f = io.open("deobfuscated_lua54.lua", "r")
if not f then
    log("ERROR: Could not open deobfuscated_lua54.lua")
    os.exit(1)
end
local vm_code = f:read("*a")
f:close()
log("-- Loaded " .. #vm_code .. " bytes")

local vm_loader, err = load(vm_code, "vm")
if not vm_loader then
    log("ERROR loading VM: " .. tostring(err))
    os.exit(1)
end

log("-- Executing VM loader...")
local VM = vm_loader()

if type(VM) ~= "table" then
    log("ERROR: VM result is " .. type(VM))
    os.exit(1)
end

log("-- Got VM table")
log("")

-- List VM keys
log("-- VM table structure:")
for k, v in pairs(VM) do
    log("  " .. tostring(k) .. " = " .. type(v))
end
log("")

-- Find the E8 function (main interpreter creator)
local E8 = VM.E8
if not E8 then
    log("ERROR: E8 not found")
    os.exit(1)
end

log("-- Found E8 function, hooking...")

-- Hook E8 to capture bytecode
VM.E8 = function(z, V, B, L)
    log("")
    log("-- E8 called!")
    log("--   z: " .. type(z))
    log("--   V: " .. type(V))

    -- Call original E8
    local ok, err = pcall(E8, z, V, B, L)
    if not ok then
        log("-- E8 error: " .. tostring(err))
    end

    -- Hook the interpreter at V[40]
    local interp = V[40] or V[0x28]
    if interp then
        log("-- Found interpreter, hooking...")

        local orig_interp = interp
        V[40] = function(W, y, I)
            func_count = func_count + 1
            local fid = func_count

            log("")
            log("-- ================================================")
            log("-- FUNCTION PROTOTYPE " .. fid)
            log("-- ================================================")

            if type(W) == "table" then
                log("-- W table has " .. #W .. " array entries")

                -- Capture all W entries
                for i = 1, 15 do
                    local v = W[i]
                    if v ~= nil then
                        if type(v) == "table" then
                            local desc = ""
                            if i == 4 then desc = "operand_A (N)" end
                            if i == 5 then desc = "operand_C (_)" end
                            if i == 7 then desc = "operand_B (d)" end
                            if i == 8 then desc = "constants_m" end
                            if i == 9 then desc = "constants_U" end
                            if i == 10 then desc = "INSTRUCTIONS (q)" end
                            if i == 11 then desc = "constants_S" end
                            log("-- W[" .. i .. "]: table with " .. #v .. " entries  " .. desc)
                            dump_array(v, "W" .. fid .. "_" .. i .. "_" .. desc:gsub("[^%w]", "_"), 3000)
                        else
                            log("-- W[" .. i .. "]: " .. type(v) .. " = " .. tostring(v))
                        end
                    end
                end
            end

            log("-- Calling original interpreter...")
            local results = {pcall(orig_interp, W, y, I)}
            if results[1] then
                log("-- Interpreter returned normally")
                return table.unpack(results, 2)
            else
                log("-- Interpreter error: " .. tostring(results[2]))
                error(results[2])
            end
        end
        V[0x28] = V[40]
    end
end

-- Try to run the VM
log("-- Looking for entry point...")

local wS = VM.wS
if wS then
    log("-- Found wS, attempting to execute...")
    local ok, err = pcall(function()
        VM:wS()
    end)
    if not ok then
        log("-- Execution error: " .. tostring(err))
    else
        log("-- Execution completed")
    end
else
    log("-- wS not found")

    -- Try to find and call any entry function
    for k, v in pairs(VM) do
        if type(v) == "function" and type(k) == "string" then
            log("-- Found function: " .. k)
        end
    end
end

log("")
log("-- ================================================")
log("-- CAPTURE COMPLETE")
log("-- Captured " .. func_count .. " function prototypes")
log("-- ================================================")

output:close()
print("\nData saved to captured_bytecode.lua")

File: capture_bytecode.py
#!/usr/bin/env python3
"""
Capture bytecode at runtime by hooking into the E8 interpreter.
"""

import re

# Read the original script
with open('/home/user/deobluraph/script.lua', 'r', encoding='utf-8', errors='replace') as f:
    content = f.read()

# Remove getgenv and comments
content = re.sub(r'^getgenv\(\)[^\n]*\n', '', content)
while content.strip().startswith('--'):
    newline_idx = content.find('\n')
    if newline_idx > 0:
        content = content[newline_idx+1:]
    else:
        break
content = content.strip()

# Create bytecode capture hooks
capture_prefix = '''-- BYTECODE CAPTURE HOOKS
print("=== BYTECODE CAPTURE ===")

-- Roblox stubs
game = { GetService = function() return {} end, PlaceId = 0 }
workspace = {}
script = { Parent = {} }
Instance = { new = function() return {} end }
getgenv = function() return _G end
gethui = function() return {} end
hookfunction = function(f) return f end
getrawmetatable = function() return {} end
setreadonly = function() end
isreadonly = function() return false end
islclosure = function() return true end
iscclosure = function() return false end
getinfo = function() return {} end
getconstants = function() return {} end
getupvalues = function() return {} end
setupvalue = function() end
getfenv = function() return _G end
setfenv = function() end
syn = nil
Drawing = nil
Input = nil

local _orig_setmetatable = setmetatable
local _orig_pairs = pairs
local _orig_ipairs = ipairs
local _orig_type = type
local _orig_tostring = tostring
local _orig_pcall = pcall
local _orig_loadstring = loadstring
local _orig_select = select

local _func_id = 0
local _captured_funcs = {}

local function dump_table(t, name, depth, max_entries)
    depth = depth or 0
    max_entries = max_entries or 100
    if depth > 3 then return end

    local indent = string.rep("  ", depth)
    local count = 0

    for k, v in _orig_pairs(t) do
        count = count + 1
        if count > max_entries then
            print(indent .. "... (more entries)")
            break
        end

        local kstr = _orig_tostring(k)
        local vtype = _orig_type(v)

        if vtype == "table" then
            local tc = 0
            for _ in _orig_pairs(v) do tc = tc + 1 end
            print(indent .. "[" .. kstr .. "] = <table:" .. tc .. ">")
            if depth < 2 and tc < 50 then
                dump_table(v, name .. "[" .. kstr .. "]", depth + 1, 30)
            end
        elseif vtype == "number" then
            print(indent .. "[" .. kstr .. "] = " .. v .. " (0x" .. string.format("%X", math.floor(v)) .. ")")
        elseif vtype == "string" then
            local s = v
            if #s > 60 then s = s:sub(1, 60) .. "..." end
            s = s:gsub("\\n", "\\\\n"):gsub("\\r", "\\\\r"):gsub("[\\x00-\\x1f]", ".")
            print(indent .. "[" .. kstr .. '] = "' .. s .. '"')
        elseif vtype == "function" then
            print(indent .. "[" .. kstr .. "] = <function>")
        else
            print(indent .. "[" .. kstr .. "] = " .. _orig_tostring(v))
        end
    end
end

-- Hook loadstring to capture and modify VM code
loadstring = function(code, name, ...)
    if _orig_type(code) ~= "string" then
        return _orig_loadstring(code, name, ...)
    end

    if #code < 1000 then
        return _orig_loadstring(code, name, ...)
    end

    print("\\n=== INTERCEPTED LOADSTRING ===")
    print("Name: " .. _orig_tostring(name))
    print("Length: " .. #code)

    -- Look for E8 function definition and inject hooks
    local modified = code

    -- Find and wrap the E8 function
    local e8_pattern = "E8=function"
    local e8_pos = modified:find(e8_pattern)

    if e8_pos then
        print("Found E8 at position " .. e8_pos)

        -- Inject hook code before E8
        local hook_code = [[
-- INJECTED BYTECODE CAPTURE HOOK
local _orig_E8_func = nil
local _capture_E8 = function(z, V, B, L)
    print("\\n=== E8 CALLED ===")

    -- V is the VM state table
    print("V type: " .. type(V))
    if type(V) == "table" then
        print("V keys:")
        for k, vv in pairs(V) do
            print("  [" .. tostring(k) .. "] = " .. type(vv))
        end
    end

    -- Original E8 creates V[0x28] = interpreter function
    -- We want to wrap that

    local result = _orig_E8_func(z, V, B, L)

    -- Now V[0x28] should be the interpreter
    if type(V) == "table" and type(V[0x28]) == "function" then
        local orig_interp = V[0x28]
        V[0x28] = function(W, y, I)
            print("\\n=== INTERPRETER CALLED ===")
            print("W (function data):")

            -- W[1] = instructions, W[5] = opA, W[10] = opcodes, etc.
            if type(W) == "table" then
                for k, v in pairs(W) do
                    local vtype = type(v)
                    if vtype == "table" then
                        local c = 0
                        for _ in pairs(v) do c = c + 1 end
                        print("  W[" .. tostring(k) .. "] = <table:" .. c .. ">")

                        -- Dump arrays that are likely bytecode
                        if type(k) == "number" and c > 0 and c < 500 then
                            print("    Contents:")
                            local cnt = 0
                            for kk, vv in pairs(v) do
                                cnt = cnt + 1
                                if cnt > 50 then
                                    print("      ... (more)")
                                    break
                                end
                                if type(vv) == "number" then
                                    print("      [" .. tostring(kk) .. "] = " .. vv .. " (0x" .. string.format("%X", math.floor(vv)) .. ")")
                                elseif type(vv) == "string" then
                                    local s = vv
                                    if #s > 40 then s = s:sub(1, 40) .. "..." end
                                    s = s:gsub("[\\x00-\\x1f]", ".")
                                    print("      [" .. tostring(kk) .. '] = "' .. s .. '"')
                                end
                            end
                        end
                    elseif vtype == "number" then
                        print("  W[" .. tostring(k) .. "] = " .. v)
                    end
                end
            end

            return orig_interp(W, y, I)
        end
        V[40] = V[0x28]  -- 0x28 = 40
    end

    return result
end
]]
        -- We'll need to replace E8=function with our wrapper
        -- This is complex - let's try a simpler approach
    end

    return _orig_loadstring(modified, name, ...)
end

print("=== EXECUTING SCRIPT ===")

local _result = (function(...)
'''

capture_suffix = '''
end)(...)

print("\\n=== EXECUTION COMPLETE ===")
print("Result type: " .. _orig_type(_result))
'''

final_script = capture_prefix + content + capture_suffix

with open('/tmp/bytecode_capture.lua', 'w') as f:
    f.write(final_script)

print(f"Bytecode capture script written: {len(final_script)} bytes")
